### ISSUE TITLE
Can't get Interceptor to work with NativeFunction calls

### ISSUE THREAD

**Issue by flop01** [frida/frida#166]
Hi,

I'd like to Intercept calls to a Win32 API that are made inside a function which I instrument with a NativeFunction object.
However, the Interceptor callback is only called when the original function call happens in the program, but it is never called when I make additional calls with my NativeFunction object.

Here is a PoC code to show what I'm trying to do.
This program waits for the user to press a key, then it calls call_systemtime() which in turn calls the GetSystemTime API:

``` c
#include <stdio.h>
#include <stdlib.h>
#include <Windows.h>

void call_systemtime(int test)
{
    SYSTEMTIME stime;

    GetSystemTime(&stime);
    printf("Systemtime called, secs = %i\n", stime.wSecond);
}

int main(void)
{
    printf("Press any key to continue\n");
    getchar();
    call_systemtime(1);
    return 0;
}
```

So, here, I want to use the Interceptor to catch calls to GetSystemTime, and I want to call the function call_systemtime() several times in my instrumentation.

The following python script does this. It attaches to the running test program and attaches the Interceptor to GetSystemTime. Then, when the user presses a key in the python script, the js code receives a message and proceeds to create a NativeFunction to call_systemtime, before calling in five times.

``` python
import frida
import sys, time

session = frida.attach("testprog.exe")

script = session.create_script("""
var gettime = Module.findExportByName("kernel32.dll", "GetSystemTime");
send("hooking GetSystemTime at " + gettime.toString());

Interceptor.attach(ptr(gettime), {
    onEnter: function(args) {
        send("entering GetSystemTime");
    }
});

var op = recv('input', function(value) {
    send("creating ptr to func");
    var dga = new NativeFunction(ptr('0x401340'), 'void', ['int']);
    for (var i = 0; i < 5; i++) {
       dga(i);
    }
});
""")

def on_message(message, data):
    print(message['payload'])

script.on('message', on_message)
script.load()

sys.stdin.read(1)
script.post_message({'type': 'input', 'payload': 'lestgo'})
sys.stdin.read()
```

This is what happens with these samples:

| User Action | Observed Behaviour |
| --- | --- |
| Run testprog.exe | runs until the getchar call and waits |
| Run python script | receive message "hooking GetSystemTime at 0x7c80176f" |
| Press enter in python script | receive "creating ptr to func", 5 calls are made to call_systemtime, program console shows 5 times "Systemtime called, secs = 55", **but** the Interceptor callback is not called |
| Press enter in program window | Normal execution of program resumes, Interceptor callback is reached and script receives "entering GetSystemTime" |

I don't get it why the original call to GetSystemTime from the program triggers the callback, but calls to my NativeFunction object don't.

Then I tried to attach OllyDbg before pressing enter in the python script. So, this is after Interceptor.attach is called but before the NativeFunction object is created and called.
I can indeed confirm that the JMP patch from frida was written at the address of GetSystemTime. If I step into this JMP, it reaches frida-agent-32 module's code, so it looks good, but still the callback doesn't send the expected string.

I'm testing this on Windows XP SP3 and Frida 7.0.11 from pip. (Not using a more recent version because I've run into other problems with them, see my other ticket https://github.com/frida/frida/issues/165)

Am I doing something wrong here ?
best regards

**Comment by oleavr** [comment]
Fixed in latest Frida 9.x.

**Comment by mrgreywater** [comment]
Doesn't work in Frida 9.1.6

**Comment by oleavr** [comment]
This is actually not a bug. Calls from Frida's own threads are ignored. Otherwise you'd be hooking functions and seeing calls that are made by Frida's internals, and that would be confusing. You can however use `Interceptor.replace()` if you need a hook that triggers regardless of caller.

**Comment by mrgreywater** [comment]
I your suggested Interceptor.replace, but NativeCallback appears to not properly clean the stack with the 'thiscall' abi on return on X86 (Windows). I'll investigate and create another bug report then.

**Comment by oleavr** [comment]
@mrgreywater Interesting. Maybe you want `stdcall`? (If you want `NativeCallback` to clear the stack, and not the caller.)

**Comment by mrgreywater** [comment]
thiscall has the first arg in ecx, and the callee cleans the stack

**Comment by oleavr** [comment]
@mrgreywater I know, but I suspect libffi implements the GCC version of `thiscall`, not the MSVC version. I'll be offline for a bit so can't investigate, but you can find the exact code we're using here: https://github.com/frida/libffi/tree/windows

**Comment by mrgreywater** [comment]
Thats it then, thanks. GCC thiscall pushes the first argument on the stack aswell. Though I guess custom calling conventions would be a nice feature for the future. Which would also help when a function expects certain arguments in xmm registers.
I assume your fastcall implementation is from msvc though?

**Comment by oleavr** [comment]
Ah ok. Btw, you can use `Memory.alloc(Process.pageSize)` (as allocations >= page size are allocated in raw pages), write your code there, then `Memory.protect()` to make it executable, and voila, you have your own callback adapter. :) (Maybe someday we should expose our code writer and relocators to JS to make this kind of trickery even easier.)

**Comment by mrgreywater** [comment]
Ok, the problem is that NativeCallback actually cleans to many arguments.

`var func = new NativeFunction(func_ptr, 'pointer', ['pointer', 'pointer','pointer', 'pointer','pointer', 'pointer'] , 'thiscall')

Interceptor.replace(func_ptr, new NativeCallback(function (ecx, stack1, stack2, stack3, stack4, stack5) {
	return func(ecx, stack1, stack2, stack3, stack4, stack5)
}, 'pointer', ['pointer', 'pointer', 'pointer','pointer', 'pointer','pointer'], 'thiscall'));
`

The NativeCallback cleans 0x18 Bytes from the stack instead of the supposed 0x14. Same goes for fastcall.

**Comment by oleavr** [comment]
This is now finally fixed in Frida 14.2.3, and libffi patches have been submitted upstream 🎉 Thanks a lot for reporting and helping narrow this one down!

**Comment by ChaosData** [comment]
This doesn't seem to work for `NativeFunction` calls that originate from an `Interceptor.attach` hook function, even if the the address of the `NativeFunction` (or of a function called by the native code invoked by the `NativeFunction` call) was `Interceptor.replace`d.

```C
#include <unistd.h>
#include <stdio.h>

extern void do_sleep();

void do_sleep() {
  sleep(1);
}

void do_something_else() {
  printf("%s\n", "hello, world");
}

int main() {
  do_sleep();
  do_something_else();
  return 0;
}
```

```JavaScript
let sleepPtr = Module.getExportByName('', 'sleep');
let _sleep = new NativeFunction(sleepPtr, 'uint', ['uint']);
Interceptor.replace(sleepPtr, new NativeCallback((seconds) => {
  console.log("my sleep called");
  return _sleep(seconds);
}, 'uint', ['uint']));

let sym_to_addr_map = {};
let addr_to_sym_map = {};

let prog = null;

Process.enumerateModulesSync().forEach(function(m) {
  if (m.name.indexOf("replace-test") != -1) {
    prog = m;
  }
});

var syms = prog.enumerateSymbols()
for (var i = 0; i < syms.length ; i++) {
  sym_to_addr_map[syms[i].name] = syms[i];
  addr_to_sym_map[syms[i].address] = syms[i];
}

let do_sleepPtr = sym_to_addr_map['do_sleep'].address;
let _do_sleep = new NativeFunction(do_sleepPtr, 'void', []);

Interceptor.attach(sym_to_addr_map['main'].address, function(args) {
  console.log("main hit, try _do_sleep");
  _do_sleep();
  console.log("did _do_sleep work?");
})
```

```
# gcc -std=c11 -o replace-test replace-test.c
# ~/.local/bin/frida --file ./replace-test -l replace-test.js --runtime v8
     ____
    / _  |   Frida 14.2.12 - A world-class dynamic instrumentation toolkit
   | (_| |
    > _  |   Commands:
   /_/ |_|       help      -> Displays the help system
   . . . .       object?   -> Display information about 'object'
   . . . .       exit/quit -> Exit
   . . . .
   . . . .   More info at https://www.frida.re/docs/home/
Spawned `./replace-test`. Use %resume to let the main thread start executing!
[Local::replace-test]-> %resume
[Local::replace-test]-> main hit, try _do_sleep
did _do_sleep work?
my sleep called
hello, world
Process terminated
```

It only seems to work when the caller of the `NativeFunction` itself was `Interceptor.replace`d:

```JavaScript
/*
Interceptor.attach(sym_to_addr_map['main'].address, function(args) {
  console.log("main hit, try _do_sleep");
  _do_sleep();
  console.log("did _do_sleep work?");
})
*/
let _main = new NativeFunction(sym_to_addr_map['main'].address, 'int', []);
Interceptor.replace(_main, new NativeCallback(() => {
  console.log("main hit, try _do_sleep");
  _do_sleep();
  console.log("did _do_sleep work?");
  return _main();
}, 'int', []));
```

```
# ~/.local/bin/frida --file ./replace-test -l replace-test-mod.js --runtime v8
     ____
    / _  |   Frida 14.2.12 - A world-class dynamic instrumentation toolkit
   | (_| |
    > _  |   Commands:
   /_/ |_|       help      -> Displays the help system
   . . . .       object?   -> Display information about 'object'
   . . . .       exit/quit -> Exit
   . . . .
   . . . .   More info at https://www.frida.re/docs/home/
Spawned `./replace-test`. Use %resume to let the main thread start executing!
[Local::replace-test]-> %resume
[Local::replace-test]-> main hit, try _do_sleep
my sleep called
did _do_sleep work?
my sleep called
hello, world
Process terminated
```
